webapp-django/crashstats/api/views.py

```diff
@@ -114,7 +116,53 @@ def is_valid_model_class(model):
     )
 
 
+def clear_empty_session(fun):
+    """Clears empty sessions so they don't persist"""
+
+    @wraps(fun)
+    def _clear_empty_session(request, *args, **kwargs):
+        # Run the function
+        ret = fun(request, *args, **kwargs)
+
+        # Clear the session if nothing was saved to it
+        session_keys = [
+            key for key in request.session.keys() if not key.startswith("_auth_user")
+        ]
+        if not session_keys:
+            request.session.clear()
+            request.session.flush()
+        return ret
+
+    return _clear_empty_session
+
+
+def no_csrf_i_mean_it(fun):
+    """Removes any csrf bits from request
+
+    This removes any csrf bookkeeping by middleware from the request so that it doesn't
+    get persisted in cookies and elsewhere.
+
+    Note: If we ever change ANON_ALWAYS setting for django-session-csrf, then we can nix
+    this. Otherwise django-session-csrf *always* creates an anoncsrf cookie regardless
+    of the existence of the anonymous_csrf_exempt decorator.
+
+    """
+
+    @wraps(fun)
+    def _no_csrf(request, *args, **kwargs):
+        ret = fun(request, *args, **kwargs)
+        # Remove any csrf bits from Django or django-session-csrf so they don't persist
+        if hasattr(request, "_anon_csrf_key"):
+            del request._anon_csrf_key
```

Comment:
What's going on here is that we have `@anonymous_csrf_exempt`, but that only exempts CHECKING the csrf--not doing the bookkeeping and adding a cookie. This `_anon_csrf_key` gets set by `process_request` in the middleware before the view code runs. So we remove it after the view runs and then the middleware doesn't add the cookie in `process_response`.